finalitygadget/finalitygadget.go

```diff
@@ -385,81 +389,83 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
+	for height := fg.lastProcessedHeight + 1; height <= latestFinalizedHeight; height++ {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			block, err := fg.queryBlockByHeight(int64(height))
+			if err != nil {
+				return fmt.Errorf("error getting block at height %d: %w", height, err)
+			}
 
-	return nil
-}
+			// only process blocks after the btc staking is activated
+			if block.BlockTimestamp < btcStakingActivatedTimestamp {
+				fg.logger.Info("Skipping block before BTC staking activation", zap.Uint64("block_height", block.BlockHeight))
+				fg.lastProcessedHeight = block.BlockHeight
+				continue
+			}
 
-func (fg *FinalityGadget) handleBlock(block *types.Block) {
-	// while block is not finalized, recheck if block is finalized every `retryInterval` seconds
-	// if finalized, store the block in DB and set the last finalized block
-	for {
-		// Check if block is finalized
-		isFinal, err := fg.QueryIsBlockBabylonFinalized(block)
-		if err != nil {
-			fg.logger.Fatal("Error checking block", zap.Uint64("block_height", block.BlockHeight), zap.Error(err))
-			return
-		}
-		if isFinal {
-			err = fg.InsertBlock(block)
+			// Query local DB for last block processed
+			localBlock, err := fg.db.QueryLatestFinalizedBlock()
 			if err != nil {
-				fg.logger.Fatal("Error storing block", zap.Uint64("block_height", block.BlockHeight), zap.Error(err))
+				return fmt.Errorf("error getting latest block from db: %v", err)
 			}
-			return
-		}
 
-		// Sleep for `PollInterval` seconds
-		time.Sleep(fg.pollInterval * time.Second)
+			// store the first babylon finalized block to local DB
+			if localBlock == nil || localBlock.BlockHeight < block.BlockHeight {
+				// Check the block is babylon finalized using sdk client
+				isFinal, err := fg.QueryIsBlockBabylonFinalized(block)
+				if err != nil && !errors.Is(err, types.ErrBtcStakingNotActivated) {
+					return fmt.Errorf("error checking block %d: %v", block.BlockHeight, err)
+				}
+				// If not finalized, throw error
+				if !isFinal {
+					return fmt.Errorf("block %d should be finalized according to client but is not", block.BlockHeight)
+				}
+
+				// If finalised, store the block in DB
+				err = fg.InsertBlock(block)
+				if err != nil {
+					return fmt.Errorf("error storing block %d: %v", block.BlockHeight, err)
+				}
+				fg.lastProcessedHeight = block.BlockHeight
+				fg.logger.Info("Inserted new finalized block", zap.Uint64("block_height", block.BlockHeight))
+			}
+		}
 	}
+
+	return nil
 }
 
 func normalizeBlockHash(hash string) string {
 	return common.HexToHash(hash).Hex()
 }
+
+func initializeLastProcessedHeight(l2Client IEthL2Client, db db.IDatabaseHandler) (uint64, error) {
+	finalizedBlock, err := l2Client.HeaderByNumber(context.Background(), big.NewInt(ethrpc.FinalizedBlockNumber.Int64()))
+	if err != nil {
+		return 0, fmt.Errorf("failed to query latest finalized block: %w", err)
+	}
+	lastProcessedHeight := finalizedBlock.Number.Uint64()
```

Comment:
on initial startup, if db is empty, it seems this condition will set `lastProcessedHeight` = `finalizedBlock` even though it is never actually processed and stored in our db. do we need to process the block if `latestBlock` is nil?

Reply:
remove this condition and only set the `lastProcessedHeight` when `latestBlock` is not nil